lms/djangoapps/courseware/entrance_exams.py

```diff
@@ -0,0 +1,175 @@
+"""
+This file contains all entrance exam related utils/logic.
+"""
+from django.conf import settings
+
+from courseware.access import has_access
+from courseware.model_data import FieldDataCache
+from courseware.models import StudentModule
+from opaque_keys.edx.keys import UsageKey
+from student.models import EntranceExamConfiguration
+from util.milestones_helpers import get_required_content
+from util.module_utils import yield_dynamic_descriptor_descendents
+from xmodule.modulestore.django import modulestore
+
+
+def feature_is_enabled():
+    """
+    Checks to see if the Entrance Exams feature is enabled
+    Use this operation instead of checking the feature flag all over the place
+    """
+    return settings.FEATURES.get('ENTRANCE_EXAMS', False)
+
+
+def course_has_entrance_exam(course):
+    """
+    Checks to see if a course is properly configured for an entrance exam
+    """
+    if not feature_is_enabled():
+        return False
+    if not course.entrance_exam_enabled:
+        return False
+    if not course.entrance_exam_id:
+        return False
+    return True
+
+
+def user_can_skip_entrance_exam(request, user, course):
+    """
+    Checks all of the various override conditions for a user to skip an entrance exam
+    Begin by short-circuiting if the course does not have an entrance exam
+    """
+    if not course_has_entrance_exam(course):
+        return True
+    if user.is_anonymous():
+        return False
+    if not user.is_authenticated():
```

Comment:
It's actually not redundant -- after removing the is_anonymous() check I
observed several tests failing in other areas of the code which were
testing anonymous users, specifically.  Adding the is_anonymous() check
back to these workflows fixed those particular tests.

So it appears there's a difference between an anonymous user, and a known
user who is simply not authenticated.

On Tue, Mar 31, 2015 at 9:39 AM, Daniel Friedman <notifications@github.com>
wrote:

> In lms/djangoapps/courseware/entrance_exams.py
> <https://github.com/edx/edx-platform/pull/7293#discussion_r27478959>:
>
> > +        return False
> > +    if not course.entrance_exam_id:
> > +        return False
> > +    return True
> > +
> > +
> > +def user_can_skip_entrance_exam(request, user, course):
> > +    """
> > +    Checks all of the various override conditions for a user to skip an entrance exam
> > +    Begin by short-circuiting if the course does not have an entrance exam
> > +    """
> > +    if not course_has_entrance_exam(course):
> > +        return True
> > +    if user.is_anonymous():
> > +        return False
> > +    if not user.is_authenticated():
>
> This check is now redundant with the previous - I think this is why you
> were seeing this line as not being covered.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/7293/files#r27478959>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org